Assignment3/Q1.py

```diff
@@ -0,0 +1,116 @@
+# coding=utf-8
+"""The approach taken in this solution is based on a DFS-like search. The worst case running time
+   is θ((NXM)^2) (where N is the number of rows and M number of columns in the given gird)but because of the optimization
+   used with isPrefix(string) method the average running time should be a lot better.
+   The solution given below performs the DFS-like search on the given grid (using words_search_recursive method)
+   but it searches ALL possible paths from each entry in a grid (corresponding to a letter).
+   This is similar to creating a big search tree with all possible words. Using isPrefix(string) method
+   it 'cuts off' early all the branches of the search tree which is being created that cannot form words."""
+
+from Dictionary import Dictionary
+
+# An array to keep track of adjacent entries in a grid.
+_ADJACENT_DELTAS = [(-1,-1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]
+
+
+def is_not_visited(i, j, rows, columns, visited):
+    """A method checks if an entry with coordinates (i,j) is contained in a grid and whether it is not yet visited.
+
+        Args:
+            i: an integer, first coordinate of an entry
+            j: an integer, second coordinate of an entry
+            rows: an integer, number of rows in a grid
+            columns: an integer, number of columns in a grid
+            visited: 2d boolean array, indicates which entries in a grid has already been visited (they are set to True)
+
+        Returns:
+            a boolean, True if an entry is in a grid and has not been visited yet, False otherwise."""
+    return 0 <= i < rows and 0 <= j < columns and not visited[i][j]
+
+
+def word_search(grid, dictionary):
+    """A method performs a word search on a character grid and returns a list of words (no duplicates) that can
+       be formed from grid's characters and are in a dictionary. Word search is case insensitive.
+
+        Args:
+            grid: a 2D array of characters, we are searching for words from dictionary in a grid
+            dictionary: a Dictionary object, contains the words to search for in a grid
+
+        Returns:
+            a set of strings (all uppercase), complete set of words that can be formed from grid's characters and
+            are in a dictionary, there are no duplicated words in a set.
```

Comment:
The phrase `there are no duplicated words in a set` is a bit redundant since a `set` cannot contain duplicates.